vendor/bat-native-ledger/src/bat/ledger/internal/publisher/publisher.cc

```diff
@@ -795,5 +792,104 @@ void Publisher::OnServerPublisherInfoLoaded(
   callback(std::move(server_info));
 }
 
+void Publisher::UpdateMediaDuration(
+    const uint64_t window_id,
+    const std::string& publisher_key,
+    uint64_t duration) {
+  BLOG(1, "Media duration: " << duration);
+  ledger_->database()->GetPublisherInfo(publisher_key,
+      std::bind(&Publisher::OnGetPublisherInfoForUpdateMediaDuration,
+                this,
+                _1,
+                _2,
+                window_id,
+                duration));
+}
+
+void Publisher::OnGetPublisherInfoForUpdateMediaDuration(
+    type::Result result,
+    type::PublisherInfoPtr info,
+    const uint64_t window_id,
+    uint64_t duration) {
+  if (result != type::Result::LEDGER_OK) {
+    BLOG(0, "Failed to retrieve publisher info while updating media duration");
+    return;
+  }
+
+  type::VisitData visit_data;
+  visit_data.name = info->name;
+  visit_data.url = info->url;
+  visit_data.provider = info->provider;
+  visit_data.favicon_url = info->favicon_url;
+
+  SaveVideoVisit(
+      info->id,
+      visit_data,
+      duration,
+      window_id,
+      [=](type::Result, type::PublisherInfoPtr) {});
+}
+
+void Publisher::GetPublisherPanelInfo(
+    const std::string& publisher_key,
+    ledger::GetPublisherInfoCallback callback) {
+  auto filter = CreateActivityFilter(
+      publisher_key,
+      type::ExcludeFilter::FILTER_ALL,
+      false,
+      ledger_->state()->GetReconcileStamp(),
+      true,
+      false);
+
+  ledger_->database()->GetPanelPublisherInfo(std::move(filter),
+      std::bind(&Publisher::OnGetPanelPublisherInfo,
+                this,
+                _1,
+                _2,
+                callback));
+}
+
+void Publisher::OnGetPanelPublisherInfo(
+    const type::Result result,
+    type::PublisherInfoPtr info,
+    ledger::GetPublisherInfoCallback callback) {
+  if (result != type::Result::LEDGER_OK) {
+    BLOG(0, "Failed to retrieve panel publisher info");
+    callback(result, nullptr);
+    return;
+  }
+
+  callback(result, std::move(info));
+}
+
+void Publisher::SavePublisherInfo(
+    const uint64_t window_id,
+    type::PublisherInfoPtr publisher_info,
+    ledger::ResultCallback callback) {
+  if (!publisher_info || publisher_info->id.empty()) {
+    BLOG(0, "Publisher key is missing for url");
+    callback(type::Result::LEDGER_ERROR);
+    return;
+  }
+
+  type::VisitData visit_data;
+  visit_data.provider = publisher_info->provider;
+  visit_data.name = publisher_info->name;
+  visit_data.url = publisher_info->url;
+  if (!publisher_info->favicon_url.empty()) {
+    visit_data.favicon_url = publisher_info->favicon_url;
+  }
+
+  SaveVideoVisit(
+      publisher_info->id,
+      visit_data,
+      0,
+      window_id,
+      [=](auto result, type::PublisherInfoPtr publisher_info) {
```

Comment:
We should generally use explicit capture lists (`[callback]` here), to prevent accidental copying.

Reply:
Fixed.